mlx/c/error.cpp

```diff
@@ -0,0 +1,45 @@
+/* Copyright © 2023-2024 Apple Inc. */
+
+#include "mlx/c/error.h"
+
+#include <cstdarg>
+#include <cstdio>
+#include <cstdlib>
+
+static void mlx_error_handler_default_(const char* msg, void* data) {
+  printf("MLX error: %s\n", msg);
+  exit(-1);
+}
+
+static void* mlx_error_handler_data_ = nullptr;
+static mlx_error_handler_func mlx_error_handler_ = mlx_error_handler_default_;
+
+extern "C" void mlx_set_error_handler(
```

Comment:
So if you wanted to have something that would allow you to "catch" the exceptions you would install something like:

```
void handler(const char* msg, void* data) {
    thread_local_storage.copy(msg);
}

mlx_set_error_handler(handler);
```

and then use it like:

```swift
if let result = mlx_array_add(ctx1, ctx2, stream.ctx) {
    return MLXArray(result)
} else {
    throw MLXArrayError(thread_local_storage.get())
}
```

or something along those lines.

If we were to adopt this (the catch/throw) in swift, we would need _all_ call sites to handle it, even if most of them just wanted the fatal error -- only the caller knows what they want to do with the error.  You could potentially build something in the handler that would allow the caller to set thread local storage that would allow you to set a per-call handler if you wanted it to default to print/fatal for most cases but wanted to catch a few.

I think this works fine for the swift side.

We might need a wrapper object for the handler + data (like the mlx_closure) unless you see this as a one-shot call and we don't need to worry about the lifecycle.

Reply:
Yes exactly!
I was seeing it as a one-shot call, but easy to add a dtor callback (see last commit)